File: paxy/interceptor/interceptor.py
from __future__ import annotations

import time

from paxy.rule.rule import Action, MatchContext, Modification, RuleManager
from paxy.store.models import Entry
from paxy.store.scope import ScopeManager
from paxy.store.store import Store


class Interceptor:
    def __init__(
        self,
        rules: RuleManager,
        store: Store,
        scope: ScopeManager | None = None,
    ) -> None:
        self._rules = rules
        self._store = store
        self._scope = scope

    def process_request(
        self,
        method: str,
        scheme: str,
        host: str,
        path: str,
        query: str,
        headers: dict[str, list[str]],
        body: bytes,
    ) -> tuple[Entry, bool]:
        # Scope check — skip recording if host is out of scope
        if self._scope and not self._scope.is_in_scope(host):
            entry = Entry(
                method=method,
                scheme=scheme,
                host=host,
                path=path,
                query=query,
                req_headers=dict(headers),
                req_body=body,
                protocol=scheme,
            )
            return entry, False  # pass through without recording

        entry = Entry(
            method=method,
            scheme=scheme,
            host=host,
            path=path,
            query=query,
            req_headers=dict(headers),
            req_body=body,
            protocol=scheme,
        )

        # GraphQL detection
        from paxy.graphql.detector import (
            extract_operation_name,
            extract_operation_type,
            is_graphql,
            parse_operation,
        )

        if is_graphql(entry):
            op = parse_operation(body)
            entry.graphql_op_type = extract_operation_type(op["query"])
            entry.graphql_operation = extract_operation_name(op["query"]) or op.get(
                "operationName", ""
            )
            entry.tags.append("graphql")
            entry.protocol = "graphql"

        ctx = MatchContext(
            method=method,
            host=host,
            path=path,
            headers=headers,
            body=body,
        )

        rule = self._rules.match(ctx)
        blocked = False

        if rule:
            if rule.action == Action.BLOCK:
                entry.tags.append("blocked")
                blocked = True
            elif rule.action == Action.MODIFY:
                headers, body = _apply_request_mods(headers, body, rule.modifications)
                entry.req_headers = dict(headers)
                entry.req_body = body
                entry.modified = True
            elif rule.action == Action.REDIRECT:
                entry.tags.append("redirected")

        self._store.add(entry)
        return entry, blocked

    def process_response(
        self,
        entry: Entry,
        status_code: int,
        headers: dict[str, list[str]],
        body: bytes,
        start_time: float,
    ) -> tuple[dict[str, list[str]], bytes]:
        entry.status_code = status_code
        entry.resp_headers = dict(headers)
        entry.resp_body = body
        entry.duration_ms = int((time.monotonic() - start_time) * 1000)

        ctx = MatchContext(
            method=entry.method,
            host=entry.host,
            path=entry.path,
            headers=headers,
            body=body,
        )

        rule = self._rules.match(ctx)
        if rule and rule.action == Action.MODIFY:
            headers, body = _apply_response_mods(headers, body, rule.modifications)
            entry.resp_headers = dict(headers)
            entry.resp_body = body
            entry.modified = True

        self._store.update(entry)
        return headers, body


def _apply_request_mods(
    headers: dict[str, list[str]],
    body: bytes,
    mods: list[Modification],
) -> tuple[dict[str, list[str]], bytes]:
    headers = dict(headers)
    for m in mods:
        if m.target == "req_header":
            headers = _apply_header_mod(headers, m)
        elif m.target == "req_body" and m.operation == "replace":
            body = m.value.encode()
    return headers, body


def _apply_response_mods(
    headers: dict[str, list[str]],
    body: bytes,
    mods: list[Modification],
) -> tuple[dict[str, list[str]], bytes]:
    headers = dict(headers)
    for m in mods:
        if m.target == "resp_header":
            headers = _apply_header_mod(headers, m)
        elif m.target == "resp_body":
            if m.operation == "replace":
                body = m.value.encode()
            elif m.operation == "find_replace":
                body = body.replace(m.find.encode(), m.replace.encode())
    return headers, body


def _apply_header_mod(headers: dict[str, list[str]], m: Modification) -> dict[str, list[str]]:
    h = {k.lower(): v for k, v in headers.items()}
    key = m.key.lower()
    if m.operation == "set":
        h[key] = [m.value]
    elif m.operation == "delete":
        h.pop(key, None)
    elif m.operation == "append":
        h.setdefault(key, []).append(m.value)
    return h

File: internal/interceptor/interceptor.go
package interceptor

import (
	"bytes"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/ykus4/paxy/internal/rule"
	"github.com/ykus4/paxy/internal/store"
)

// Interceptor wires the rule engine into the HTTP flow and records traffic.
type Interceptor struct {
	Rules *rule.Manager
	Store *store.Store
}

func New(rules *rule.Manager, st *store.Store) *Interceptor {
	return &Interceptor{Rules: rules, Store: st}
}

// ProcessRequest applies rules to an outgoing request and records it.
// Returns (entry, block) where block=true means the request should be dropped.
func (ic *Interceptor) ProcessRequest(req *http.Request, scheme string) (*store.Entry, bool) {
	body, _ := io.ReadAll(req.Body)
	req.Body = io.NopCloser(bytes.NewReader(body))

	entry := &store.Entry{
		Method:    req.Method,
		Scheme:    scheme,
		Host:      req.Host,
		Path:      req.URL.Path,
		Query:     req.URL.RawQuery,
		ReqHeader: cloneHeader(req.Header),
		ReqBody:   body,
		Protocol:  scheme,
	}

	ctx := &rule.MatchContext{
		Method:  req.Method,
		Host:    req.Host,
		Path:    req.URL.Path,
		Headers: req.Header,
		Body:    body,
	}

	if r := ic.Rules.Match(ctx); r != nil {
		switch r.Action {
		case rule.ActionBlock:
			entry.Tags = append(entry.Tags, "blocked")
			ic.Store.Add(entry)
			return entry, true
		case rule.ActionModify:
			applyRequestMods(req, r.Modifications)
			entry.Modified = true
			newBody, _ := io.ReadAll(req.Body)
			req.Body = io.NopCloser(bytes.NewReader(newBody))
			entry.ReqBody = newBody
			entry.ReqHeader = cloneHeader(req.Header)
		case rule.ActionRedirect:
			entry.Tags = append(entry.Tags, "redirected")
		}
	}

	ic.Store.Add(entry)
	return entry, false
}

// ProcessResponse records the response and applies modifications.
func (ic *Interceptor) ProcessResponse(entry *store.Entry, resp *http.Response, start time.Time) {
	body, _ := io.ReadAll(resp.Body)
	resp.Body = io.NopCloser(bytes.NewReader(body))

	entry.StatusCode = resp.StatusCode
	entry.RespHeader = cloneHeader(resp.Header)
	entry.RespBody = body
	entry.DurationMs = time.Since(start).Milliseconds()

	ctx := &rule.MatchContext{
		Method:  entry.Method,
		Host:    entry.Host,
		Path:    entry.Path,
		Headers: resp.Header,
		Body:    body,
	}

	if r := ic.Rules.Match(ctx); r != nil && r.Action == rule.ActionModify {
		applyResponseMods(resp, r.Modifications)
		newBody, _ := io.ReadAll(resp.Body)
		resp.Body = io.NopCloser(bytes.NewReader(newBody))
		entry.RespBody = newBody
		entry.RespHeader = cloneHeader(resp.Header)
		entry.Modified = true
	}

	ic.Store.Update(entry)
}

func applyRequestMods(req *http.Request, mods []rule.Modification) {
	for _, m := range mods {
		if m.Target != "req_header" && m.Target != "req_body" {
			continue
		}
		switch m.Target {
		case "req_header":
			applyHeaderMod(req.Header, m)
		case "req_body":
			if m.Operation == "replace" {
				req.Body = io.NopCloser(strings.NewReader(m.Value))
			}
		}
	}
}

func applyResponseMods(resp *http.Response, mods []rule.Modification) {
	for _, m := range mods {
		switch m.Target {
		case "resp_header":
			applyHeaderMod(resp.Header, m)
		case "resp_body":
			if m.Operation == "replace" {
				resp.Body = io.NopCloser(strings.NewReader(m.Value))
			} else if m.Operation == "find_replace" {
				body, _ := io.ReadAll(resp.Body)
				newBody := strings.ReplaceAll(string(body), m.Find, m.Replace)
				resp.Body = io.NopCloser(strings.NewReader(newBody))
			}
		}
	}
}

func applyHeaderMod(h http.Header, m rule.Modification) {
	switch m.Operation {
	case "set":
		h.Set(m.Key, m.Value)
	case "delete":
		h.Del(m.Key)
	case "append":
		h.Add(m.Key, m.Value)
	}
}

func cloneHeader(h http.Header) map[string][]string {
	out := make(map[string][]string, len(h))
	for k, vs := range h {
		cp := make([]string, len(vs))
		copy(cp, vs)
		out[k] = cp
	}
	return out
}
